File: python/discount_applier.py
from dataclasses import dataclass


@dataclass
class User:
    name : str


class Notifier:
    def notify(self, user: User, message: str):
        raise RuntimeError("Can't call this from a unit test, it will make a network call")


class DiscountApplier:
    def __init__(self, notifier: Notifier):
        self.notifier = notifier

    def apply_v1(self, discount: float, users: list[User]):
        # Bug! Should be `range(0, len(users))`
        for i in range(1, len(users)):
            message = f"You've got a new discount of {discount}%"
            user = users[i]
            self.notifier.notify(user, message)

    def apply_v2(self, discount: float, users: list[User]):
        for i in range(0, len(users)):
            message = f"You've got a new discount of {discount}%"
            # Bug! Should be `users[i]`
            user = users[0]
            self.notifier.notify(user, message)

File: ts/discount-applier.ts
import {Notifier} from "./notifier";

export class DiscountApplier {
  private notifier: Notifier;

  constructor (notifier: Notifier) {
    this.notifier = notifier
  }

  applyV1 (discount: number, users: string[]) {
    // Bug! Should be `let i = 0`
    for (let i = 1; i < users.length; i++) {
      const message = `You've got a new discount of ${discount}`
      const user = users[i]
      this.notifier.notify(user, message)
    }
  }

  applyV2 (discount: number, users: string[]) {
    for (let i = 0; i < users.length; i++) {
      const message = `You've got a new discount of ${discount}`
      // Bug! Should be `users[i]`
      const user = users[0]
      this.notifier.notify(user, message)
    }
  }
}
